### ISSUE TITLE
[S138] queue exceeded!!

### ISSUE THREAD

**Issue by guindous** [scouter-project/scouter#294]
We have being using Scouter 0.4.18 for months with no problems. Recently we noted that after weeks of server (collector) running produced high CPU usage and logs showed repeatedly (every 10-11 seconds):
`[S138] queue exceeded!!`
After restart the server problems go away. Now the work around is to restart the server after some days running.

In order to add more info about logs, the first time that the message appeared was after a NullPointerException. Part of this log:

```
20161021 08:54:55 [S206] Auto deletion... 20161012
20161021 09:25:11 [S159] invalid data
java.lang.NullPointerException
        java.io.OutputStream.write(OutputStream.java:75)
        scouter.server.netio.data.MultiPacket$$anonfun$toBytes$1.apply$mcVI$sp(MultiPacket.scala:51)
        scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:141)
        scouter.server.netio.data.MultiPacket.toBytes(MultiPacket.scala:50)
        scouter.server.netio.data.MultiPacketProcessor$.add(MultiPacketProcessor.scala:64)
        scouter.server.netio.data.NetDataProcessor$.processCafeMTU(NetDataProcessor.scala:107)
        scouter.server.netio.data.NetDataProcessor$.process(NetDataProcessor.scala:88)
        scouter.server.netio.data.NetDataProcessor$$anonfun$1$$anonfun$apply$1.apply$mcV$sp(NetDataProcessor.scala:66)
        scouter.server.util.ThreadScala$$anon$1.run(ThreadScala.scala:12)
20161021 10:49:28 [S138] queue exceeded!!

```

Any hint about why this could happens? Anyone with similar symptoms? 

Thanks in advance.

**Comment by gunlee01** [comment]
Is the user CPU of the scouter process high?
If processing is delayed and more than a certain number of queues accumulate, the Scouter is designed to empty it in order not to cause a system failure.
There are various reasons for processing delays, so it is difficult to tell.
If this problem persists, please create a thread dump at that time dozens of times at intervals of several seconds and share it.

**Comment by guindous** [comment]
@gunlee01 thanks for your fast support. I ran this bash script to obtain thread dumps of collector java pid during high processing and "queue exceeded!!" log message:

```for (( c=1; c<=12; c++)); do jstack {pid} > threaddump.`date +%H_%M_%S`_$c; sleep 8; done;```

There is the result for download: https://www.dropbox.com/s/5usb54my02d2yfd/scouter_td.zip?dl=0

If you need any extra information don't hesitate to ask.

Thanks!

**Comment by gunlee01** [comment]
All `scouter.server.db.TextPermWR` thread stack is at `KeyFile.getTimeKey()` of your dump files.
It's very strange I've never seen.

Carefully guessing

1. This problem does not seem to be a disk I/O problem, but check the server's disk I/O.

2. Lack of heap memory and broken index file get worse it.
 - Try to increase heap memory on the scouter start script. (`-Xmx512m -> -Xmx1024m`)
 - and backup your data files and delete it.(`scouter.server/database/*`)
3. another reason maybe.
 - You can also use multiple servers for better performance.
   (Scouter client can connect multiple servers in an user view.)

**Comment by guindous** [comment]
@gunlee01 Following your guessing:

1. Not I/O problem. Even if processor is high, I/O is calm.
2. I have added more heap memory to the Server. If problem reappears, I will backup and delete data file. I will proceed in order to have a precise answer of what produces this issue.
3. That is not possible. That server is collecting traffic from only one tomcat instance. A VERY heavily used one.

I will update you with next tests related with item 2.
Thanks for your support.

**Comment by guindous** [comment]
In order to close this issue, I share the tests related last post's item 2. First, after I duplicated heap memory server problem reappeared. So, didn't solve the issue. After that, I backuped and deleted data files and the problem disappeared. Because I suspected the data corruption problem where related with auto deletion process, I waited at least 2 weeks (mgr_purge_keep_days=14) and auto deletion process worked ok and everything is running smoothly so far.

The only mystery left is to understand why data index got corrupted.

Thanks for your support and help.
